File: ecommerce-services/order-service/src/main/java/com/ecommerce/order/service/OrderService.java
package com.ecommerce.order.service;

import com.ecommerce.order.dto.OrderRequest;
import com.ecommerce.order.dto.OrderResponse;
import com.ecommerce.order.entity.Order;

import java.util.List;

public interface OrderService {
    
    /**
     * Create new order and start saga
     */
    OrderResponse createOrder(OrderRequest request);
    
    /**
     * Get order by ID
     */
    OrderResponse getOrderById(Long id);
    
    /**
     * Get order by order number
     */
    OrderResponse getOrderByOrderNumber(String orderNumber);
    
    /**
     * Get all orders for a user
     */
    List<OrderResponse> getUserOrders(Long userId);
    
    /**
     * Get orders by user and status
     */
    List<OrderResponse> getUserOrdersByStatus(Long userId, Order.OrderStatus status);
    
    /**
     * Cancel order (user-initiated)
     */
    OrderResponse cancelOrder(Long id, String reason);
    
    /**
     * Get order status
     */
    Order.OrderStatus getOrderStatus(Long id);
    
    /**
     * Get saga status (for debugging)
     */
    Order.SagaStatus getSagaStatus(Long id);
    
    /**
     * Get all orders (admin)
     */
    List<OrderResponse> getAllOrders();
    
    /**
     * Get failed orders (monitoring)
     */
    List<OrderResponse> getFailedOrders();
}

File: ecommerce-frontend/src/app/core/services/order.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { OrderRequest, OrderResponse, OrderStatus } from '../models/order.model';
import { ApiResponse } from '../models/auth.model';

@Injectable({
  providedIn: 'root'
})
export class OrderService {
  private readonly API_URL = 'http://localhost:8084/api/orders'; // Order service port
  private http = inject(HttpClient);

  /**
   * Create new order
   * POST /api/orders
   */
  createOrder(request: OrderRequest): Observable<OrderResponse> {
    return this.http.post<ApiResponse<OrderResponse>>(this.API_URL, request)
      .pipe(map(response => response.data));
  }

  /**
   * Get order by ID
   * GET /api/orders/{id}
   */
  getOrderById(id: number): Observable<OrderResponse> {
    return this.http.get<ApiResponse<OrderResponse>>(`${this.API_URL}/${id}`)
      .pipe(map(response => response.data));
  }

  /**
   * Get order by order number
   * GET /api/orders/number/{orderNumber}
   */
  getOrderByOrderNumber(orderNumber: string): Observable<OrderResponse> {
    return this.http.get<ApiResponse<OrderResponse>>(`${this.API_URL}/number/${orderNumber}`)
      .pipe(map(response => response.data));
  }

  /**
   * Get all orders for a user
   * GET /api/orders/user/{userId}
   */
  getUserOrders(userId: number): Observable<OrderResponse[]> {
    return this.http.get<ApiResponse<OrderResponse[]>>(`${this.API_URL}/user/${userId}`)
      .pipe(map(response => response.data));
  }

  /**
   * Get orders by user and status
   * GET /api/orders/user/{userId}/status/{status}
   */
  getUserOrdersByStatus(userId: number, status: OrderStatus): Observable<OrderResponse[]> {
    return this.http.get<ApiResponse<OrderResponse[]>>(
      `${this.API_URL}/user/${userId}/status/${status}`
    ).pipe(map(response => response.data));
  }

  /**
   * Cancel order
   * PUT /api/orders/{id}/cancel
   */
  cancelOrder(id: number, reason?: string): Observable<OrderResponse> {
    let params = new HttpParams();
    if (reason) {
      params = params.set('reason', reason);
    }
    
    return this.http.put<ApiResponse<OrderResponse>>(
      `${this.API_URL}/${id}/cancel`,
      null,
      { params }
    ).pipe(map(response => response.data));
  }

  /**
   * Get order status
   * GET /api/orders/{id}/status
   */
  getOrderStatus(id: number): Observable<string> {
    return this.http.get<ApiResponse<string>>(`${this.API_URL}/${id}/status`)
      .pipe(map(response => response.data));
  }

  /**
   * Get saga status (for debugging)
   * GET /api/orders/{id}/saga-status
   */
  getSagaStatus(id: number): Observable<string> {
    return this.http.get<ApiResponse<string>>(`${this.API_URL}/${id}/saga-status`)
      .pipe(map(response => response.data));
  }

  /**
   * Get all orders (admin)
   * GET /api/orders
   */
  getAllOrders(): Observable<OrderResponse[]> {
    return this.http.get<ApiResponse<OrderResponse[]>>(this.API_URL)
      .pipe(map(response => response.data));
  }

  /**
   * Get failed orders (monitoring)
   * GET /api/orders/failed
   */
  getFailedOrders(): Observable<OrderResponse[]> {
    return this.http.get<ApiResponse<OrderResponse[]>>(`${this.API_URL}/failed`)
      .pipe(map(response => response.data));
  }
}
